Sources/PenguinGraphs/ApproximateKNN.swift

```diff
@@ -48,28 +92,40 @@ extension BidirectionalGraph {
   {
     var workList = workList  // Make mutable.
     for seed in seeds {
-      workList.push(seed, at: distanceBetween(query, seed, &self))
+      vertexVisitationState.set(seed, in: &self, to: .gray)
+      workList.push(seed, at: similarityBetween(query, seed, &self))
     }
-    while true {
+    var closestK = [(VertexId, Similarity)]()
+    while closestK.count < k {
       guard let nearest = workList.top else { fatalError("No items in work list.") }
-      if vertexVisitationState.get(nearest, in: self) == .white {
+      if vertexVisitationState.get(nearest, in: self) == .gray {
         for e in edges(from: nearest) {
           let neighbor = destination(of: e)
-          let distance = distanceBetween(query, neighbor, &self)
-          workList.push(neighbor, at: distance)
+          if neighbor == query || vertexVisitationState.get(neighbor, in: self) != .white {
+            continue
+          }
+          vertexVisitationState.set(neighbor, in: &self, to: .gray)
+          let similarity = similarityBetween(query, neighbor, &self)
+          workList.push(neighbor, at: similarity)
         }
         for e in edges(to: nearest) {
           let neighbor = source(of: e)
-          let distance = distanceBetween(query, neighbor, &self)
-          workList.push(neighbor, at: distance)
+          if neighbor == query || vertexVisitationState.get(neighbor, in: self) != .white {
+            continue
+          }
+          vertexVisitationState.set(neighbor, in: &self, to: .gray)
+          let similarity = similarityBetween(query, neighbor, &self)
+          workList.push(neighbor, at: similarity)
         }
         vertexVisitationState.set(nearest, in: &self, to: .black)
       } else {
-        // We've seen the same nearest vertex, so we're done here!
-        workList.heap.sort()
-        return workList.prefix(k).map { ($0.payload, $0.priority) }
+        assert(vertexVisitationState.get(nearest, in: self) == .black)
+        // We've seen the same nearest vertex, so pop it off.
+        let nearest = workList.pop()!
+        closestK.append((nearest.payload, nearest.priority))
```

Comment:
It seems like the change above in the while loop conditions means we could find the same vertex again since we've popped it off the heap (e.g. it's the neighbor of the next highest similarity). I think previous code (while loop as just true and returning the sorted top K) was correct, but maybe I'm missing something.

Reply:
Hehe, as it turns out, I ran into exactly that scenario in testing! This code actually handles that correctly because we only add vertices to the `workList` when they're colored `.white`. So we never re-add them if they're `.gray` or `.black`.